apps/yggdrasil_web/web/controllers/character_controller.ex

```diff
@@ -64,4 +86,8 @@ defmodule YggdrasilWeb.CharacterController do
         render conn, :errors, data: err_changeset
     end
   end
+
+  defp admin?(conn) do
+    EnsurePermission.all?(conn, character: [:admin])
```

Comment:
So admin probably should have been called `:all` instead. So the basic idea is similar to a file system in some respects so like the following (replacing `:admin` with `:all`)
```
:read,  :all -> you could list all characters and show any character
:write, :all -> you can modify, delete and create any character
:read        -> can only view your characters
:write       -> you can modify, delete, and create only characters for your self.
```
I limited it to three, which is less granular, so as you can see if you have `:write` you can create, update and delete. That said of you course you could add in `:update`, `:delete` for finer control.

The obvious limitation to this is that you can't selectively apply all, in that I could not grant a user the ability to see all characters but only allow them to create, modify or delete their own.

The question is that do we truly need that level of control? If we do then perhaps a set like this would be needed
```
:create
:create_all
:read
:read_all
:update
:update_all
:delete
:delete_all
```
the only issue here is ambiguity in that do you need both or just one in the controllers? I think the idiom should be both `:read` and `:read_all`. `:read` gets you into the resource and `read_all` augments your abilities within that action. Easy enough to define idioms of course, just pointing it out.

with this scheme we would get something like
```elixir
plug EnsurePermission, [character: [:read]] when action in [:index, :show]

def index(...) do
   if EnsurePermission.all?(conn, character: [:read_all]) do
     # show all 
   else
     # show only ones for user
   end
end
```
where `:read_all` doesn't say you have access but rather modifies your access.

I did read your notes about composition btw, just using this as an example. I'm somewhat partial to the first one, but if we truly need more granularity I'm not opposed to this.